src/sync_user.hpp

```diff
@@ -0,0 +1,118 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_SYNC_USER_HPP
+#define REALM_OS_SYNC_USER_HPP
+
+#include <string>
+#include <memory>
+#include <unordered_map>
+
+namespace realm {
+
+struct SyncSession;
+
+// A `SyncUser` represents a single user account. Each user manages the sessions that
+// are associated with it.
+class SyncUser {
+public:	
+	enum class State {
+		LoggedOut, Active, Error;
```

Comment:
I think we should take a step back. If the app logs in with a username + incorrect password and the ROS is functioning properly, the auth server should reject that login request. The binding will then be responsible for asking the user to provide a different password and try again. In this case the binding won't even get as far as creating a `realm::SyncUser` for its binding-level User object.

The only way the `User` can get into the error state is if the server is broken and the auth and sync parts of the ROS are not configured properly. In this case I'm not sure what the best way to recover is, or if we should even attempt to recover.